### ISSUE TITLE
✨ Create AWS Data Sync Instance 

### ISSUE THREAD

**Issue by darren1988** [ministryofjustice/analytical-platform#5175]
### Describe the context.

We embarked on this originally earlier in the year, where the request came in for a datasync instance that would allow OPG to move various pieces of unstructured/semi-structured data (PDFs, Documents etc.) into the Analytical Platform, so that they could be accessed directly from the AP without having to download files from a fileshare and manually reupload them. This would allow the data to be automatically replicated to our account from the fileshare, meaning analysts would be able to natively access all their files. This was for a good while pending the creation of a service account from ATOS, but said account has been created. 

Work required:

We need to create an AWS Datasync Instance, and set it up to connect to/authenticate with the fileshare, using the service account provided by ATOS

### Definition of done

- [x] Modernisation Platform's shared VPC added to analytical-platform-ingestion
- [x] Data Sync instance and agent deployed
- [x] Data Sync location configured
- [x] Data Sync destination configured
- [x] Tested

**Comment by YvanMOJdigital** [comment]
requires defined architecture before planning

**Comment by darren1988** [comment]
Meeting scheduled for 29/08/24 to discuss scope and technical architecture for this work

**Comment by bagg3rs** [comment]
I have the service account credentials from Gwion and have put them into 1Password `OPG - AWS DataSync Service Account` AP Shared Account

**Comment by darren1988** [comment]
To be discussed at refinement.

**Comment by jacobwoffenden** [comment]
Reached out to @ministryofjustice/modernisation-platform about adding their shared platform VPC into our ingestion account

**Comment by jacobwoffenden** [comment]
Have agreed with @ministryofjustice/modernisation-platform that this isn't a problem, we can add shared VPC, will inspect environment code in modernisation-platform and modernisation-platform-environments

**Comment by jacobwoffenden** [comment]
shared VPC added to ingestion account, *however* upon further reading Data Sync *does not* support shared VPCs

**Comment by jacobwoffenden** [comment]
Plan is to create VPCs using existing, soon to be retired, never connected to MoJ TGW, ranges from MP

**Comment by jacobwoffenden** [comment]
VPC build-out in progress, EC2 instance build-out also in progress.

However the DataSync registration **is not** programatic, the DataSync server needs to be accessible from whatever machine is running Terraform/registering manually in the console. This is problematic because

1) GitHub Actions is our primary CI/CD system that has no internal connectivity to our VPC and isn't really in scope for making work because its MP's system
1) Our VPC currently has no public connectivity
1) From what I've read, the AWS provided AMI does not include SSM agent, instead need you need to connect via SSH (I tried serial console but the default of `admin` / `password` didn't work)
1) The above presents a chicken/egg problem 🤔 
  
Do I open the endpoint to GitHub Actions? GlobalProtect?

Do I add userdata to install SSM agent and write the activation key to Secrets Manager? I don't even know if the activation key is held on disk or if I'd to run a command...

**Comment by jacobwoffenden** [comment]
10/10/24 update:
- Instance up
- Able to route using NLB, however when Terraform is triggering activation, the target group becomes unhealthy presumably because DataSync activation server stops, it just hangs, need to possibly add `private_link_endpoint`

**Comment by jacobwoffenden** [comment]
16/10/24 update:

- Call with AWS who pointed out we were missing an egress rule on the VPC endpoint security group and the ENI security group needs attaching to the VPC endpoint, don't think either of those things are documented 🤷 
- DataSync agent is now registered and as expected, the agents server stops so the NLB is just sat

TODO:
- [x] Attach VPCs to MoJ TGW
- [x] Add credentials to Secrets Manager
- [x] Add DataSync location - can we look up dom1 by IP? 
- [ ] Add DataSync task
- [x] Add S3 destination bucket
- [x] Decide where data will be replicated to

**Comment by jacobwoffenden** [comment]
Currently blocked by https://github.com/ministryofjustice/modernisation-platform/issues/8275

**Comment by darren1988** [comment]
Requested support from mod platform to help unblock this ticket

**Comment by jacobwoffenden** [comment]
NVVS/LAN&Wifi team have given me access to https://github.com/ministryofjustice/deployment-tgw, so I'm not as blocked as last week 🙏

**Comment by jacobwoffenden** [comment]
Moving back to blocked pending information on connecting to DOM1 from AWS

**Comment by jacobwoffenden** [comment]
Blocked while https://github.com/ministryofjustice/modernisation-platform/pull/8322 is deployed

**Comment by jacobwoffenden** [comment]
31/10/24 update (spooky edition 🎃):

- Connected to MoJO debug instance via SSM
- Cannot resolve dom1.infra.int...
    ```bash
    [root@ip-10-26-128-43 bin]# dig dom1.infra.int
    ;; communications error to 10.26.128.2#53: timed out
    ;; communications error to 10.26.128.2#53: timed out
    ;; communications error to 10.26.128.2#53: timed out
    ```
- Can resolve google.com...

**Comment by jacobwoffenden** [comment]
⬆ pull request added staff device production vpc to mod platform tgw route table, still not responding...

traffic is arriving in destination vpc, but return path from ark might not be working, need to reach out further

**Comment by jacobwoffenden** [comment]
This pull request (https://github.com/ministryofjustice/deployment-tgw/pull/258) has allowed us to lookup dom1.infra.int via MoJO DNS resolver

**Comment by jacobwoffenden** [comment]
Blocked again as we liaise with other parties about direct connection into Ark over TGW.

It can route back, we just cant route there, presumably because its blocked by a Palo Alto

**Comment by jacobwoffenden** [comment]
🎉 I am able to connect to DOM1 from my debugging instance! 🎉

**Comment by jacobwoffenden** [comment]
Reached out to ATOS because I can't access one of the locations

**Comment by jacobwoffenden** [comment]
Have reached out to @gwionap for clarification on source data

**Comment by jacobwoffenden** [comment]
Updated locations received from @gwionap, will continue.

**Comment by jacobwoffenden** [comment]
Have created a task but is failing...

![Image](https://github.com/user-attachments/assets/668b0c43-dfbd-404e-a5ea-3009a4ffead8)

I can't explore this location with smbclient from the debug instance either 

```
smb: \> ls hq/PGO/Shared/Group
do_connect: Connection to eucw4171nas002.dom1.infra.int failed (Error NT_STATUS_IO_TIMEOUT)
Unable to follow dfs referral [\eucw4171nas002.dom1.infra.int\mojshared002$]
do_list: [\hq\PGO\Shared\Group] NT_STATUS_IO_TIMEOUT
```

have escalated to @gwionap

**Comment by jacobwoffenden** [comment]
A more verbose output from smbclient

```
smb: \> ls hq/PGO/Shared/Group/
dos_clean_name [\hq\PGO\Shared\Group\]
unix_clean_name [\hq\PGO\Shared\Group\]
signed SMB2 message (sign_algo_id=1)
signed SMB2 message (sign_algo_id=1)
sitename_fetch: No stored sitename for realm ''
namecache_fetch: no entry for eucw4171nas002.dom1.infra.int#20 found.
resolve_hosts: Attempting host lookup for name eucw4171nas002.dom1.infra.int<0x20>
namecache_store: storing 1 address for eucw4171nas002.dom1.infra.int#20: 10.172.69.24
Connecting to 10.172.69.24 at port 445
convert_string_handle: E2BIG: convert_string(UTF-8,CP850): srclen=30 destlen=16 error: No more room
Connecting to 10.172.69.24 at port 139
do_connect: Connection to eucw4171nas002.dom1.infra.int failed (Error NT_STATUS_IO_TIMEOUT)
Unable to follow dfs referral [\eucw4171nas002.dom1.infra.int\mojshared002$]
do_list: [\hq\PGO\Shared\Group\] NT_STATUS_IO_TIMEOUT
signed SMB2 message (sign_algo_id=1)
signed SMB2 message (sign_algo_id=1)
signed SMB2 message (sign_algo_id=1)
signed SMB2 message (sign_algo_id=1)
signed SMB2 message (sign_algo_id=1)
```

Seeing the following in VPC flow logs

```
2 730335344807 eni-05ad72cf6a35649b0 10.26.128.43 10.172.69.24 33266 139 6 3 180 1731627738 1731627767 ACCEPT OK
```

So maybe its the routing back from ATOS?

**Comment by jacobwoffenden** [comment]
SMB traffic is being dropped at the Palo Altos 💀

**Comment by jacobwoffenden** [comment]
@bagg3rs has raised a demand with Tech Services

**Comment by bagg3rs** [comment]
Update from Demand [here](https://mojdt.slack.com/archives/C04M8224WCV/p1732875065049309?thread_ts=1732874377.001679&cid=C04M8224WCV)

**Comment by jacobwoffenden** [comment]
Things are happening in TS, just slowly.

**Comment by jacobwoffenden** [comment]
Blocked **again**, potential permissions issue on file server...

**Comment by jacobwoffenden** [comment]
Raised with ATOS using original request

**Comment by jacobwoffenden** [comment]
ATOS have asked us to raise a new request but we got around that by going through Networks

Our DataSync user was added to a new group and once we switched to testing with `mount` it worked,`smbclient` wasn't working as expected

I initiated a run yesterday, it detected 800k~ files but errored after 120k~

**Comment by jacobwoffenden** [comment]
Spoken to @gwionap and the suggestion is to scope to a specific repository

**Comment by jacobwoffenden** [comment]
Source scoped as advised by @gwionap, schedule added for 2300 on Wednesday so will run tonight

**Comment by jacobwoffenden** [comment]
scheduled run failed, a variety of errors which I cannot via screenshot, but are:

* Source location responded with: Stale file handle (content changed unexpectedly or became unavailable)

**Comment by jacobwoffenden** [comment]
I've triggered a new task with changes to reporting and verification https://github.com/ministryofjustice/modernisation-platform-environments/pull/8658/commits/fdd8c9eebc1f7ee18c91da60e63b9c0753b3d667

**Comment by jacobwoffenden** [comment]
Still failed, but did transfer some data.

Pausing until new year, we have a catch up with OPG.

**Comment by jacobwoffenden** [comment]
Awaiting list of Excel databases to fetch

**Comment by jacobwoffenden** [comment]
Liaised with OPG, have new location to fetch

**Comment by jacobwoffenden** [comment]
That also failed, have reached out to AWS via #ext-aws and support 173693416500007

**Comment by jacobwoffenden** [comment]
Extra permissions have been granted to our user 🎉 1 millions files have been transferred (took 1 day, 7 hours, 26 minutes),  however 182k were skipped 🤔 

Running again to see what happens

**Comment by jacobwoffenden** [comment]
Failed on same folder permissions

Have raised with OPG how to proceed, because we cannot engineer our way out of this

**Comment by jacobwoffenden** [comment]
TODO:

- [x] Create a secret to store exclusion list, this contains PII, so cannot be public
- [x] Use exclusion list in task
- [x] New bucket in AP (dev + prod)
- [x] Set up replication from source (AP-Ingestion) -> destination (AP-DP) (dev + prod)
- [x] Empty dev bucket / enable batch replication
- [x] Enable replication to destination dev bucket in APDP (TBC)
- [x] Run task again / kick off historic replication

**Comment by Gary-H9** [comment]
As per the above PRs:

- New buckets in AP (dev + prod):
`mojap-data-production-datasync-opg-ingress-development` ✅
`mojap-data-production-datasync-opg-ingress-production` ✅

- Set up replication from source (AP-Ingestion) -> destination (AP-DP) (dev + prod)
I've tested replication by dropping files in both development and production. ✅

A diagram of this piece of work is included below:

![Image](https://github.com/user-attachments/assets/ac7d038b-212b-41f9-9bce-ebeda346adbf)

**Comment by jacobwoffenden** [comment]
`exec-00f3592435d079670` started in production

**Comment by jacobwoffenden** [comment]
Task failed due to one file, so I'm declaring that a success.

Have updated requestor with S3 and KMS details, awaiting access test.

**Comment by jacobwoffenden** [comment]
Chased with requestor, they are going to set up this datasource in database access to test

- https://github.com/moj-analytical-services/data-engineering-database-access/pull/5040

**Comment by jacobwoffenden** [comment]
Have updated SMB hostname after previous one was shut down without notice (at least to us). New one also failed with "host is down" after 3 hours 🤷 

Also have added "*/~$*" to try and cut any temp files from the filesystem

**Comment by darren1988** [comment]
@darren1988 to contact tech service to find a contact for file storage

**Comment by jacobwoffenden** [comment]
Have reached out to TA for file and storage

**Comment by jacobwoffenden** [comment]
Limited response from EUCS

**Comment by DaleyDelivery** [comment]
Darren to reach out

**Comment by jacobwoffenden** [comment]
Palo Alto kit is flagging this as suspicious, has been escalated to DISO.

**Comment by DaleyDelivery** [comment]
DISO will investigate further during their next sprint

**Comment by jacobwoffenden** [comment]
Exclude secret updated and DataSync config updated via Terraform apply.

**Comment by jacobwoffenden** [comment]
🎉 last DataSync task was successful
<img width="1512" alt="Image" src="https://github.com/user-attachments/assets/af1566b3-ba4e-4404-8cf8-f83356e3374a" />

I am going to close this ticket and deal with any subsequent work with the normal process
